### DISCUSSION TITLE
Caddy branch

### DISCUSSION THREAD

**Discussion by rcarmo** [piku/piku#382] [Ideas]
We might as well get started discussing this, since a) it's Summertime and I have at least a theoretical amount of (mental) bandwidth coming (even as actual bandwidth by the countryside/beach trends towards zero) and b) #381 was just filed by @seguri and it reminded me that we should get a move on if we want it done by... 2025.

So here are the things we are relying on `uwsgi` for:

* Managing environment variables exposed to apps
* Process supervision, with:
  * some nice built-in support for some runtimes (esp. Python)
  * multiple (different) workers per app
  * a pretty great worker logging functionality (not brilliant, but stupefyingly useful)
  * automatic reload of those workers when the respective configuration file changes (which is most of the secret sauce of why `piku` is so simple)
* Management of socket connections:
  * Directly via WSGI (since `uwsgi` runs the system Python interpreter and loads apps directly)
  * Indirectly via TCP sockets (for everything else but Python)
  * By piping directly from the `nginx` domain socket (which is one of the reasons why WSGI apps are fast)
 * Crontabs
 * Other minor stuff that we can complete later, but which is largely non-structural

And here are the things we are relying on `nginx` to do:

* Exposing outside HTTP/HTTPS ports
* Mapping those ports to domains and apps
* Serving things with the right TLS certificates (with a dependency on ACME for public certificates, and a relatively sane fallback for self-signed)
* Doing gzip compression
* Doing static file serving
* Doing content caching (by storing some request results on disk)
* Likely something else I've missed 

Caddy does _many_ of those things and has the _potential_ to simplify a few more. However, there are some caveats:

* It has only recently been added to conservative distributions like Debian (please adjust your interpretation of "recent" according to the geological time scales Debian appears to operate at)
* We would need to stick to the packaged version (currently 2.6.2-5 on Debian, as an example - 0.2 versions behind latest) 
* Some of the functionality we need might only be available via plugins (which are quite likely to be packaged separately)
* Pretty much all the configuration management we do today for `uwsgi` (which is braindead simple) would have to be thrown out and rewritten from scratch
* Same for `nginx` (although there would likely be much less to re-write given that Caddy groks ACME)
* We would need to figure out supervisors, domain and TCP socket forwarding, logging, etc. (nothing insurmountable, but slow and error-prone)
* Probably a few more annoyances that I haven't thought about yet

There would also be some likely breakage--for instance, WSGI support might go out the window, although I suspect people might just move to bundle `gunicorn` or `aiohttp` or `fastapi` into their apps since the Python landscape has become quite trendy. But moving to a pure (domain or otherwise) socket-driven app model wouldn't hurt.

And, of course, the helper scripts, cloud-init templates, etc. need to be rewritten. That's "just" deployment, but it's still work.

The end goal would still be the same--rely only on OS packages, do an `apt install caddy click` and drop Piku into place.

I have given an entire five seconds to the notion of rewriting Piku in some other language, but the simplicity of it being a "tiny" Python script is largely what makes it so appealing, useful (and approachable) to many people, so I'm not really keen on doing that (and especially not a language that would require us to ship binaries, even if we could build them via CI on Github). We might try to remove `click` as a dependency, though (I've given some thought to the idea of writing a few decorators that do just what we need and keep the code simple).

If I had to pick another programming language for a full rewrite, it would be either Go or Guile Scheme, the first because it provides trivial cross-compiling and the second because I wanted to raise the blood pressure of a few people who managed to read this far and I actually considered writing the first version of Piku in some sort of Scheme :)

**Comment by chr15m** [comment]
> Scheme... raise the blood pressure

Lol 😆 

> moving to a pure (domain or otherwise) socket-driven app model wouldn't hurt

I'm in favour of this. The special treatment of Python had actually bitten me in the past. I have a client currently waiting for a reply to me re "how can we run some apps with a non-system python?"

Let me play devil's advocate for a second - if we go this route could we get away with just systemd, nginx, and a little bit of Python glue?

> Remove click...[replace with] a few decorators

Love this idea! I wonder how lean we can go with this? 🤔

**Reply by rcarmo** [reply]
Oh, you wouldn't really _need_ `nginx`. At all. I happen to like it a lot, but aiming for a full replacement through Caddy seems like something that would also lower RAM requirements, etc. -- my usual embedded system mindset at play.

I have an idea about the replacement decorators, just haven't tried it yet -- it's just wrapping stdlib argument parsing in decorators that behave like the few we use from click.

**Reply by rcarmo** [reply]
Ok, here's a little toy (replaced by a hackier version that cheats on the error messages):

```python
import sys
from functools import partial

commands = {}

def cmd(*args, name=None):
    """draft decorator for commands"""
    def inner(func):
        n = name or func.__name__
        commands[n] = {
            "cmd": func,
            "doc": func.__doc__
        }
        return func
    return inner

# Usage example
@cmd()
def hello(someone: str):
    """say hello to the user"""
    print(f'Hello, {someone}!')

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} [command] [options]")
        for cmd, data in commands.items():
            print(f"{cmd} - {data['doc']}")
    else:
        cmd = sys.argv[1]
        if cmd in commands:
            try:
                commands[cmd]["cmd"](*sys.argv[2:])
            except Exception as e:
                print(f"{e.__str__().replace("()","")}")
        else:
            print(f"Unknown command: {cmd}")
```

**Reply by seguri** [reply]
I feel like time spent replacing `click` is kind of wasted:
- we are already dependant on python
- `click` is a good library
- the plugin system works with it
- why reinventing the wheel?

Of course doing it for fun is a perfectly fine motivation, but there might be items with higher priorities first :)

**Reply by rcarmo** [reply]
Actually, I would like to remove `click` because we've been caught out by some distributions shipping a very old version of it. It's happened more than once over the past eight years. The plugin system is something I entirely forgot about, because, TBH, we don't have that many plugins.

**Comment by rcarmo** [comment]
I've picked this up lately and have a semi-working prototype with `systemd` replacing `uwsgi` and `caddy` working locally. Will probably push something to a separate repo soon, once I get the real bonus feature (`podman`) working properly.

**Reply by chr15m** [reply]
Very cool, keen to check it out.

**Comment by chr15m** [comment]
Something I've been toying with is an even leaner script without special cases for each deploy type that just expects a shell script in the repo for the deploy step. So e.g. to deploy a python app you'd have a script that does 'pip install -r requirements.txt' instead of piku doing it for you. Maybe a bit too minimal.

**Reply by seguri** [reply]
I did the same. I installed a new server, got used to Caddy, deployed all the apps under `/opt/apps`, and wrote a bash script very similar to piku, e.g. `apps release <app> <version>`. And instead of trying to adapt the script to every possible runtime out there, my script just defines an interface and leaves the implementation to the app itself: pre-release.sh, install.sh, release.sh, etc.

**Reply by rcarmo** [reply]
Yeah, that's why I'm considering ditching every single other runtime except Python (which we need anyway) and just pack the rest inside a container, which will be built as you release. Podman is just because it's easy to integrate with systemd.

**Comment by rcarmo** [comment]
Okay, so here's a brief status report of a little experimentation:

- I can setup Caddy via the API endpoint without any real issues
- I can also set up workers as user systemd services. It is a colossal pain, but doable
- I have given up on Podman because setting up rootless podman and quadlets is a maze of paper cuts, especially inside the LXC development environments I use

What I did learn is that trying to do the entire thing with Profile + ENV + Caddy configuration is just too many moving parts, so I sort of reset the approach and right now I'm doing a little side trip into all-out Docker Compose integration.

One YAML file with both Docker and Caddy configurations (not the `caddy-docker` plugin, an actual Caddy file embedded in the compose, one single source of truth for all app settings, a very minor set of sensible defaults for storage and paths.

If I can get that to work nicely, then I'll probably publish that first. If there's any interest in the systemd approach _now_, then I can put up a gist of the current draft, but it isn't feature complete nor clean.

**Comment by chr15m** [comment]
@rcarmo what's the value proposition here of switching to Docker? I guess I'm wondering why somebody wouldn't just use one of the mature Docker based self-hosted PaaS solutions. It seems to me a lot of people run Piku because it doesn't rely on Docker.

> If there's any interest in the systemd approach now, then I can put up a gist of the current draft, but it isn't feature complete nor clean.

This sounds very useful. If you have time to share I'd be interested!

> I can also set up workers as user systemd services. It is a colossal pain, but doable

Would love to know more about this. In my head it should simply be systemd keeping some processes alive but you've obviously got real experience in this now destroying that fantasy. :sweat_smile:

Thanks for pursuing this, seems like valuable work.

**Reply by jfmatth** [reply]
Agree, if you go docker than you're just another dokku which we don't want.

Why is Caddy so hard?  I thought you'd be able to replace new UWSGI endpoints with Caddy ones?

is there a branch of this i can play with?

**Reply by rcarmo** [reply]
> @rcarmo what's the value proposition here of switching to Docker? I guess I'm wondering why somebody wouldn't just use one of the mature Docker based self-hosted PaaS solutions. It seems to me a lot of people run Piku because it doesn't rely on Docker.

I am running many services using Portainer these days and wanted something simpler that would manage both kinds of applications. That’s why I am interested in Podman. But since Podman doesn’t work well in the environments I use, I can reason out how to use containers faster with Docker. That doesn’t mean we have to use that part for Piku (although I am hoping I could later just replace docker with podman if I can sort out the nesting).

> > If there's any interest in the systemd approach now, then I can put up a gist of the current draft, but it isn't feature complete nor clean.
> 
> This sounds very useful. If you have time to share I'd be interested!

I can publish the repo and label it as completely experimental, here be dragons, etc.

> > I can also set up workers as user systemd services. It is a colossal pain, but doable
> 
> Would love to know more about this. In my head it should simply be systemd keeping some processes alive but you've obviously got real experience in this now destroying that fantasy. 😅

The thing is that replacing uwsgi with systemd unit files requires a lot of finagling. Timers are a particular pain because you have to create a .service and a .timer and convert the cron notation, etc. But the real pain is managing the environment variables and runtimes (I stripped everything else but Python to have a testable thing), because the spawn_app part has to be completely rewritten from scratch.

When I got to try to use Podman quadlets I realized that there is no easy way to reconcile the Procfile/ENV approach with that _and_ that rootless Podman is stiil very brittle, so I decided to se how far I could go with docker-compose without using the buildpack overhead of dokku, but removed the unit files to have a small testable thing. Both things are on separate branches so they’re independently reusable.

> Thanks for pursuing this, seems like valuable work.

I’m sure we can reuse the systemd portions and the Caddy portions without much hassle, although I haven’t yet tested HTTPS. I’m starting with static sites, then multiple workers, etc. - I got to the point where I can run a couple of the samples with minimal changes.

**Reply by rcarmo** [reply]
> Agree, if you go docker than you're just another dokku which we don't want.
> 
> Why is Caddy so hard? I thought you'd be able to replace new UWSGI endpoints with Caddy ones?

Caddy isn’t hard (at least as far as I have gone with it). It’s replacing uwsgi with systemd that’s been annoying, because I have to essentially pack a lot more into a unit file (and in different formats) than in a uwsgi.ini file _and_ I have to kick systemd to notice the changes (whereas uwsgi will just pick them up). 

And systemd is just… bad. Usable, but translating all of our Procfile features into it is a little maze of constant annoyance, and the logging facilities are atrocious.

Base Caddy (at least in Debian Trixie) does not have a process supervisor I can use, so I can’t replace both nginx and uwsgi with it. It also refused to include multiple Caddyfiles in the same way nginx does, so I just POST stuff into its internal API right now.

> is there a branch of this i can play with?

Soon. I am working in a separate repo that I will move to the org. I called it “kata” (for repetition) since I am currently running it alongside piku on a couple of machines and it’s currently meant to be a set of exploratory iterations, not necessarily something we can replace piku with outright.

**Reply by rcarmo** [reply]
Here you go (both branches are somewhat broken, but the code is readable): https://github.com/piku/kata

**Reply by jfmatth** [reply]
Thanks, ill have a look.

**Reply by chr15m** [reply]
@rcarmo one thing I've been experimenting with which could be useful is nix running on ubuntu/debian. It's fairly lightweight isolation without containers. I put together this very small repo demonstrating what is possible.

https://github.com/chr15m/minimal-direnv-nix

For a Piku-next we could do away with direnv (or use it?) and just use the Nix part. Deps could be specified more simply and inserted into an external file with a nicer syntax that's included by the flake.nix or something.

**Reply by chr15m** [reply]
I am not sure if anybody else is interested in this but I also find myself wishing for a solution that doesn't *require* a git push. Some other type of copy like an scp or rsync. I often do a final build on my local system and having the build checked in is sometimes not what I want (though these days I'm tending more and more towards things that don't need a build step, at least when prototyping).

**Reply by chr15m** [reply]
By the way, `nix` only requires one dep: `apt install nix-tools` and after that you can get this self contained environment. I wonder what version of Caddy is available in Nix? :thinking:

**Reply by jfmatth** [reply]
> I am not sure if anybody else is interested in this but I also find myself wishing for a solution that doesn't *require* a git push. Some other type of copy like an scp or rsync. I often do a final build on my local system and having the build checked in is sometimes not what I want (though these days I'm tending more and more towards things that don't need a build step, at least when prototyping).

I actually love the fact that it's a git push, other tools use the same technology and I'd like to stick with that

**Reply by chr15m** [reply]
> I actually love the fact that it's a git push, other tools use the same technology and I'd like to stick with that

Yeah maybe I need to build some other thing for this use case.

**Reply by jfmatth** [reply]
Does Caddy have the ability to run a program on a webhook call?  That might solve both our problems 😄

**Reply by jfmatth** [reply]
apparently it does, quick nicely too...

https://blog.dashdreams.com/blog/WebServer/Webhook-Based-On-Caddy/#4-config-caddyfile

**Reply by rcarmo** [reply]
That's a plugin that I don't actually think ships with the `caddy` package in `trixie` or `noble`, but I can check. I, too, am interested in doing some sort of webhook deployment, and if I can get it to work on `kata` that too can be re-used.

**Reply by rcarmo** [reply]
> @rcarmo one thing I've been experimenting with which could be useful is nix running on ubuntu/debian. It's fairly lightweight isolation without containers. I put together this very small repo demonstrating what is possible.

I've actually been using `nix` for a few things. It is tempting (as in Dark Side tempting). I'll take a look.

**Reply by chr15m** [reply]
Since I mentioned Nix this is an interesting & related read: 

https://blog.railway.com/p/introducing-railpack

**Reply by jfmatth** [reply]
> I am not sure if anybody else is interested in this but I also find myself wishing for a solution that doesn't _require_ a git push. Some other type of copy like an scp or rsync. I often do a final build on my local system and having the build checked in is sometimes not what I want (though these days I'm tending more and more towards things that don't need a build step, at least when prototyping).

(just getting back to reading all these discussions).

I love the ```git push...``` aspect of PIKU and did a whole talk on this.  In the end that's the flow that I want, but understand if there are others.

**Comment by rcarmo** [comment]
Meanwhile, status update: the `kata` `docker` branch now deploys NodeJS apps as well (badly, but at least they're uh... contained?) and its just 917 loc. I'm going to play with it a bit more...

**Reply by rcarmo** [reply]
Am now at 843 LOC. I realized that if I used docker swarm I got secret management for free and that if I just passed through  `docker` commands I didn't need to implement ps, logs, scale, etc., so... it keeps shrinking.

**Comment by jfmatth** [comment]
Can't recall if this was covered... https://caddy-snake.readthedocs.io/en/latest/

**Reply by rcarmo** [reply]
That's very cute. However, I haven't looked at using caddy for anything but acting as reverse proxy (my current post-piku toy uses containers extensively).

But it might be a good nginx/uwsgi replacement if it ends up being packaged in Debian.

**Reply by jfmatth** [reply]
I just played around with it and it's pretty limiting, might be worth just figuring out how to have Python answer the HTTP call directly (i.e. no W/ASGI, then caddy-snake isn't needed.
